### PR TITLE
Design & document HTTP API

### PR DESCRIPTION
**Pull request by bfirsh**
This is a work-in-progress implementation of #413. I was considering creating an issue to discuss the design of the HTTP API, but I figure a pull request and some documentation might be the most productive medium for doing this.

The text of the documentation is incomplete and might be partial sentences. Mainly looking here for feedback on the API, then we can complete all the sentences.

I think it is mostly there, but a few design questions:

1. Do we like our request and response objects?
2. Do we like our status names? (Maybe `failure` instead of `failed`? Is there some prior art here we can build upon instead of inventing something new?)
3. #428
4. Is this compatible with #437 and #443?
5. File handling is a work in progress and I will create a separate piece of documentation about that.
6. Is this compatible with a potential future async version?
7. Do we want to version it? #94

### CODE REVIEW

File: docs/http.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    docker run -d -p 5000:5000 --init --gpus all my-model
+
+    # if you're on an M1 Mac:
+    docker run -d -p 5000:5000 --init --platform=linux/amd64 my-model
+
+Port 5000 is now serving the API:
+
+    curl http://localhost:5000
+
+To view the API documentation in browser for the model that is running, open [https://localhost:5000/docs](https://localhost:5000/docs).
```
Comment (zeke) [inline docs/http.md:23]: Does `https` work here?

File: docs/http.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+## `GET /openapi.json`
+
+The [OpenAPI](https://swagger.io/specification/) specification of the API.
```
Comment (zeke) [inline docs/http.md:27]:
```suggestion
The [OpenAPI](https://swagger.io/specification/) specification of the API, which is derived from the input and output types specified in your model's [Predictor](python.md) object.
```

**Review by zeke — APPROVED** [review]
Great start! Left a few suggestions, and I will follow up with responses to your questions.

We get a `/docs` page for free too, right? But maybe we don't want to advertise that because it's ugly and probably doesn't work?

File: docs/http.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# HTTP API
```
Comment (sirupsen) [inline docs/http.md:1]: Why not inline this in the README for now for maximum discoverability? I think fat READMEs are underrated, and I think this is high level enough to be there. I always Cmd-F for stuff

File: docs/http.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+## `GET /openapi.json`
+
+The [OpenAPI](https://swagger.io/specification/) specification of the API.
+
+## `POST /predictions`
+
+Make a single prediction. The request body should be a JSON object with the following fields:
+
+- `input`: a JSON object with the same keys as the [arguments to the `predict()` function](python.md). Any `File` or `Path` inputs are passed as URLs.
```
Comment (sirupsen) [inline docs/http.md:33]: Worth a sentence about the type-magic you accomplish for these signatures, it's clever and great, but bears repeating here without reading that whole doc

File: docs/http.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Responds with:
+
+    {
+        "status": "success",
+        "output": "data:image/png;base64,..."
+    }
+
+Or, with curl:
+
+    curl -X POST -H "Content-Type: application/json" -d '{"input": {"image": "https://example.com/image.jpg", "text": "Hello world!"}}' http://localhost:5000/predictions
```
Comment (sirupsen) [inline docs/http.md:61]: Do we need to provide a timeout arg here? Often these are long and slow, no?

File: docs/http.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+Then, start the Docker container:
+
+    docker run -d -p 5000:5000 --init my-model
+
+    # If your model uses a GPU:
+    docker run -d -p 5000:5000 --init --gpus all my-model
+
+    # if you're on an M1 Mac:
+    docker run -d -p 5000:5000 --init --platform=linux/amd64 my-model
```
Comment (sirupsen) [inline docs/http.md:17]: Have you got this working with [M1's GPU yet](https://nod.ai/pytorch-m1-max-gpu/)?

**Review by sirupsen — APPROVED** [review]

File: docs/http.md (hunk 3 above)
Comment (bfirsh) [reply docs/http.md:1]: Hmm, yes. [I like long web pages too.](https://clig.dev/) We're thinking of turning this into some proper docs at some point too, so we'll have to think in context of that.

The downside of one long page is then search engines suffer. You can't google "Cog HTTP API" and get deep linked to the right place.

**Comment by bfirsh** [comment]
@sirupsen @zeke Sorry I should have been clearer in the PR body: I'm mainly looking for feedback here on the actual API. The text of this documentation is incomplete (but thank you for the suggestions!).

@zeke I see answers to the questions are incoming. 😄

File: docs/http.md (hunk 4 above)
Comment (bfirsh) [reply docs/http.md:33]: Agreed. This needs expanding.

File: docs/http.md (hunk 6 above)
Comment (bfirsh) [reply docs/http.md:17]: Unfortunately I think it's very very hard to navigate through xhyve etc.

File: docs/http.md (hunk 5 above)
Comment (bfirsh) [reply docs/http.md:61]: Yes -- good idea. We have a timeout for the Redis worker, but not HTTP.

File: docs/http.md (hunk 1 above)
Comment (bfirsh) [reply docs/http.md:23]: No, no, it does not...

File: docs/http.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+## `POST /predictions`
+
+Make a single prediction. The request body should be a JSON object with the following fields:
+
+- `input`: a JSON object with the same keys as the [arguments to the `predict()` function](python.md). Any `File` or `Path` inputs are passed as URLs.
+- `output_file_prefix`: A base URL to upload output files to. See [file handling documentation](...) for more details.
+
+The response is a JSON object with the following fields:
+
+- `status`: Either `success` or `failed`.
```
Comment (zeke) [inline docs/http.md:38]: Per your question in the PR body: `failure` seems preferable here for consistency with `success`.

File: docs/http.md (hunk 1 above)
Comment (bfirsh) [reply docs/http.md:23]: I just realized Copilot wrote this for me! Not my fault! Honest!

File: docs/http.md (hunk 6 above)
Comment (sirupsen) [reply docs/http.md:17]: I'm hoping Apple invests in it.... would make the Ultra and whatever Mac Pro is to come more appealing

Comment (bfirsh) [reply docs/http.md:17]: Yeah 100%.

It's starting to work in some form, but it only works if you run PyTorch and Tensorflow on the bare metal (and deal with dependency hell).

File: docs/http.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+When a Cog Docker image is run, it serves an HTTP API for making predictions.
+
+First, build your model:
+
+    cog build -t my-model
+
+Then, start the Docker container:
+
+    docker run -d -p 5000:5000 --init my-model
```
Comment (zeke) [inline docs/http.md:11]: Is `--init` needed here? And if so, why?

I read the [docker docs on `--init`](https://docs.docker.com/engine/reference/run/), but still don't really get what this is about.

> You can use the `--init` flag to indicate that an init process should be used as the PID 1 in the container. Specifying an init process ensures the usual responsibilities of an init system, such as reaping zombie processes, are performed inside the created container.
https://www.loom.com/share/41790390b7f1435c980bd5adddce7ff8

Comment (zeke) [reply docs/http.md:11]: https://www.loom.com/share/41790390b7f1435c980bd5adddce7ff8

**Comment by bfirsh** [comment]
For clarity for future reviewers: this is really for feedback on the API, not for feedback on the documentation. I have only half written the documentation. I'm particularly interested in the questions in the original post.

I opened this pull request instead of making a spec for the API because I figured it would be easier to comment on a PR than a GitHub issue.

This is also related to #443 where there are some ideas about how bi-directional HTTP might work. (Not necessarily for now, but could influence future design.)

File: docs/http.md (hunk 8 above)
Comment (bfirsh) [reply docs/http.md:11]: It's so signals work. Ctrl-C, etc.

**Comment by bfirsh** [comment]
I wonder how an async API would work. It could be along the lines of:

```
POST /predictions?async=1
{
  "id": "332f6f8f39f1",
  "status": "processing",
}
```

```
GET /predictions/332f6f8f39f1
{
  "id": "332f6f8f39f1",
  "status": "success",
  "output": ...
}
```

**Comment by sirupsen** [comment]
@bfirsh yeah I think something along those lines.

There's probably ~4 distinct cases to think through:

| HTTP Version | URL                      | Result                                                                                                                     |
|--------------|--------------------------|----------------------------------------------------------------------------------------------------------------------------|
| HTTP/1.1     | /predictions             | HTTP Uni-directional SSE Stream                                                                                            |
| HTTP/1.1     | /predictions?async=false | HTTP 201                                                                                                                   |
| HTTP/{2,3}   | /predictions             | [HTTP Bi-directional stream](https://developers.google.com/web/fundamentals/performance/http2#streams_messages_and_frames) |
| HTTP/{2,3}   | /predictions?async=false | HTTP 201

**Comment by zeke** [comment]
> Do we like our request and response objects?

For reference, here's an example of what those objects look like today:

```json
{
  "Request": {
    "title": "Request",
    "type": "object",
    "properties": {
      "input": {
        "$ref": "#/components/schemas/Input"
      },
      "output_file_prefix": {
        "title": "Output File Prefix",
        "type": "string"
      }
    }
  },
  "Response": {
    "title": "Response",
    "required": ["status"],
    "type": "object",
    "properties": {
      "status": {
        "$ref": "#/components/schemas/Status"
      },
      "output": {
        "$ref": "#/components/schemas/Output"
      },
      "error": {
        "title": "Error",
        "type": "string"
      }
    },
    "description": "The status of a prediction."
  }  
}
```

Two things stand out to me:

1. `title` is redundant. Let's get rid of it. https://github.com/replicate/cog/issues/457
1. `description` is inconsistent. Here's a PR to fix that: https://github.com/replicate/cog/pull/498

**Comment by evilstreak** [comment]
> Do we like our status names? (Maybe failure instead of failed? Is there some prior art here we can build upon instead of inventing something new?)

I've had a look, but can't find any sort of open standard here.

GitHub Actions use `success`, `failure` and `cancelled`, which works, but isn't exactly consistent. I think I'd prefer `succeeded`, `failed` and `cancelled` for consistency.

This random Microsoft status enum is very self-consistent: https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.jobstate?view=powershellsdk-7.0.0. I like the distinction between `…ed` statuses for a thing that's over and `…ing` statuses for something in progress. (One small nit: I'd probably prefer something like `Waiting` rather than `NotStarted`.

I think adopting `…ing` and `…ed` endings would make it easier to add consistent new statuses in the future. If we start with `success` and `failure` then we either end up inconsistent like GitHub Actions or we add a status like `cancellation` which feels super weird.

With `…ing` and `…ed`, if we want to add statuses for jobs that haven't started or are in progress, it's easy to find verbs: `queuing`, `waiting`, `running`.

File: docs/http.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        "input": {
+            "image": "https://example.com/image.jpg",
+            "text": "Hello world!"
+        }
+    }
+
+Responds with:
+
+    {
+        "status": "success",
```
Comment (evilstreak) [inline docs/http.md:55]:
```suggestion
        "status": "succeeded",
```

File: docs/http.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+## `POST /predictions`
+
+Make a single prediction. The request body should be a JSON object with the following fields:
+
+- `input`: a JSON object with the same keys as the [arguments to the `predict()` function](python.md). Any `File` or `Path` inputs are passed as URLs.
+- `output_file_prefix`: A base URL to upload output files to. See [file handling documentation](...) for more details.
+
+The response is a JSON object with the following fields:
+
+- `status`: Either `success` or `failed`.
```
Comment (evilstreak) [inline docs/http.md:38]: As outlined in https://github.com/replicate/cog/pull/475#issuecomment-1070805035, I'd prefer sticking to "ed" and "ing" endings:

```suggestion
- `status`: Either `succeeded` or `failed`.
```

File: docs/http.md (hunk 8 above)
Comment (bfirsh) [reply docs/http.md:11]: Zeke says this is fixed in FastAPI now!

**Comment by bfirsh** [comment]
Thanks for all your feedback here! I'm going to merge this as-is because it's a lot better than nothing, then turn the suggestions about the API into new tickets.

**Comment by bfirsh** [comment]
👆 here are the tickets I made

**Pull request merged**